### ISSUE TITLE
delivery acknowledgements

### ISSUE THREAD

**Issue by M-LAK** [amqp/rhea#329]
Hi, 

- I want to know if RHEA supports delivery acknowledgements ? If yes, which parameter should we modify ?
- We can make setTimeout for ACK ?
- we can save message received as file ?

Thank you

**Comment by grs** [comment]
>     * I want to know if RHEA supports delivery acknowledgements ? If yes, which parameter should we modify ?

Yes, messages are acknowledged by default (the accepted event will be triggered for a positive acknowledgement, the rejected. released or modified event will be triggered for a negative acknowledgement, i.e. when the delivery was not confirmed, and the settled event will also be triggered in both cases). 

>     * We can make setTimeout for ACK ?

There is nothing built in to timeout waiting for an acknowledgement. You would need to build that yourself (i.e. timeout if the relevant ack event has not fired when the timeout fires).

>     * we can save message received as file ?

Again thereis nothing in this library that would save a message as a file, you will need to implement that yourself.

**Comment by M-LAK** [comment]
Ok, So we can add "accepted event" in receiver API ? 

i have another question: i want to create a durable subscriber on topic, it works well, but the message still in topic after after subscription.  DS not consume message.

My code: 

_var args = require('../options.js').options({
    'client': { default: 'my-client', describe: 'name of identifier for client container'},
    'subscription': { default: 'DS', describe: 'name of identifier for subscription'},
    't': { alias: 'topic', default: 'TopicAMQP_Protocol', describe: 'name of topic to subscribe to'},
    'h': { alias: 'host', default: '******', describe: 'dns or ip name of server where you want to connect'},
    'p': { alias: 'port', default: ******, describe: 'port to connect to'}
}).help('help').argv;

var connection = require('rhea').connect({ port:args.port, host: args.host,username:'******', password: '*****', container_id:args.client });
connection.on('receiver_open', function (context) {
    console.log('subscribed');
});
connection.on('message', function (context) {
    console.log('message' + context.message);
});

// the identity of the subscriber is the combination of container id
// and link (i.e. receiver) name
connection.open_receiver({name:args.subscription, source:{address:args.topic, durable:true, expiry_policy:'never', capabilities:['topic']}});_

What is wrong ?
Thank you

**Comment by grs** [comment]
> So we can add "accepted event" in receiver API ?
The API already lets you handle the accepted event. See e.g. https://github.com/amqp/rhea/blob/master/examples/simple_send.js#L35

**Comment by grs** [comment]
For you other query, what broker are you using? I would suggest turning on protocol tracing at the broker and confirming that the library does indeed send an acknowledgement.

**Comment by M-LAK** [comment]
-Q1:  Yes, but i want the same thing in open_receiver not in open_sender.  It's possible ? 
-Q2: I use Universal Messaging (WebMethods).

**Comment by grs** [comment]
> -Q1: Yes, but i want the same thing in open_receiver not in open_sender. It's possible ?

You can add an event handler for the event, but in usual usage you will not get an event at the receiver because it is the receiver that sends the accept (to the sender). The accepted event indicates that the peer has accepted the message (i.e. has positively acknowledged it). 

> -Q2: I use Universal Messaging (WebMethods).

I'm not familiar with that. I would consult their documentation to determine how to enable protocol tracing and/or any more details on thei durable subscription support. You can also use wireshark to verify that the library is acknowledging the message (I am sure it is, since this is used all the time for many other applications and with various different brokers). Of course you can also see it from rhea logging, but that is less convincing then confirmation from the wire or the broker side.

**Comment by M-LAK** [comment]
Q1: Ok thank you, because i need to prove  that we don't lose messages if the receiver is not going well.
Q2: Ok I will check, but you confirm that the code is correct and normally DS consume message ?

**Comment by grs** [comment]
Yes it looks fine to me.

**Comment by M-LAK** [comment]
Hello,

I still have an acknowledgment problem with my durable subscriber.
I created a DS with java, and the acknowledgment works perfectly.

So what parameter must be added to my js code to activate acknowledgment?

My code:

var args = require('../options.js').options({
'client': { default: 'my-client', describe: 'name of identifier for client container'},
'subscription': { default: 'DS', describe: 'name of identifier for subscription'},
't': { alias: 'topic', default: 'TopicAMQP_Protocol', describe: 'name of topic to subscribe to'},
'h': { alias: 'host', default: '******', describe: 'dns or ip name of server where you want to connect'},
'p': { alias: 'port', default: ******, describe: 'port to connect to'}
}).help('help').argv;

var connection = require('rhea').connect({ port:args.port, host: args.host,username:'*', password: '', container_id:args.client });
connection.on('receiver_open', function (context) {
console.log('subscribed');
});
connection.on('message', function (context) {
console.log('message' + context.message);
});

// the identity of the subscriber is the combination of container id
// and link (i.e. receiver) name
connection.open_receiver({name:args.subscription, source:{address:args.topic, durable:true, expiry_policy:'never', capabilities:['topic']}});

Thank you

**Comment by grs** [comment]
Acknowledgements are enabled by default so no extra option is required. I would verify with a tcpdump or broker side logging (e.g. try setting env var  PN_TRACE_FRM=1 for broker) that the accept is indeed being sent by the client. You can then compare with what you see for your java client and/or consult with support for your broker as to why messages are not being dequeued.
